### PR TITLE
community[minor]: Add support for Azure AI Search vector store

### LINKED ISSUE #2349: Azure Cognitive Search
Is there plans to integrate Azure Cognitive Search as vector store?

### PR DESCRIPTION
**Pull request by sinedied**
This PR adds Azure AI Search vectore store integration.
This is a follow-up of https://github.com/langchain-ai/langchainjs/pull/2396 that has been stuck for months. I forked the original PR to keep the prior contribution history, I hope it's ok that way.

Many changes happened since the original PR, as Azure Cognitive Search was renamed Azure AI Search,the SDK was updated with breaking changes, and LangChainJS structure has also changed.
I updated all of these, and went a bit further to align with the existing Azure Cosmos DB integration documentation, examples and tests.

> **Note:** this is a first version to unblock the usage of this service, but I plan to push it a bit further in another PR with options to customize the index schema and fit existing indexes.

Fixes #2349 
Twitter: @sinedied

### CODE REVIEW

**Review by pablocastro — CHANGES_REQUESTED** [review]
Thanks for adding Azure AI Search support! I left a number of comments in the PR, happy to clarify anything in the comments.

File: docs/api_refs/typedoc.json [hunk 1]
```diff
@@ -235,6 +236,7 @@
     "../../langchain-core/src/utils/types.ts",
     "../../langchain-core/src/vectorstores.ts",
     "../../libs/langchain-anthropic/src/index.ts",
+    "../../libs/langchain-cloudflare/src/index.ts",
```
[the remark this answers is not in the document]
Comment (sinedied) [reply docs/api_refs/typedoc.json:239]: Not really, this file auto-generated when building Langchain so it was probably missed in a previous PR

File: docs/api_refs/typedoc.json [hunk 2]
```diff
@@ -225,6 +225,7 @@
     "../../langchain-core/src/utils/async_caller.ts",
     "../../langchain-core/src/utils/chunk_array.ts",
     "../../langchain-core/src/utils/env.ts",
+    "../../langchain-core/src/utils/function_calling.ts",
```
[the remark this answers is not in the document]
Comment (sinedied) [reply docs/api_refs/typedoc.json:228]: Same as above, this file is generated automatically

File: langchain/package.json [hunk 3]
```diff
@@ -994,6 +994,7 @@
     "@aws-sdk/client-sagemaker-runtime": "^3.310.0",
     "@aws-sdk/client-sfn": "^3.310.0",
     "@aws-sdk/credential-provider-node": "^3.388.0",
+    "@aws-sdk/signature-v4": "^3.374.0",
```
[the remark this answers is not in the document]
Comment (sinedied) [reply langchain/package.json:997]: Good catch, this comes from the previous PR and I missed that one in the cleanup

File: libs/langchain-community/src/vectorstores/azure_aisearch.ts [hunk 4]
```diff
@@ -0,0 +1,694 @@
...
+  private readonly options: AzureAISearchQueryOptions;
+
+  constructor(embeddings: EmbeddingsInterface, config: AzureAISearchConfig) {
+    super(embeddings, config);
+
+    const endpoint =
+      config.endpoint ?? getEnvironmentVariable("AZURE_AISEARCH_ENDPOINT");
+    const key = config.key ?? getEnvironmentVariable("AZURE_AISEARCH_KEY");
+
+    if (!config.client && !endpoint && !key) {
```
[the remark this answers is not in the document]
Comment (sinedied) [reply libs/langchain-community/src/vectorstores/azure_aisearch.ts:143]: Thanks, fixed

File: libs/langchain-community/src/vectorstores/azure_aisearch.ts [hunk 5]
```diff
@@ -0,0 +1,694 @@
...
+    const docsWithScore: [Document, number][] = [];
+
+    for await (const item of results) {
+      const document = new Document<
+        AzureAISearchDocumentMetadata & { embedding: number[] }
+      >({
+        pageContent: item.document[DEFAULT_FIELD_CONTENT],
+        metadata: {
+          ...item.document[DEFAULT_FIELD_METADATA],
+          embedding: item.document[DEFAULT_FIELD_CONTENT_VECTOR],
```
[the remark this answers is not in the document]
Comment (sinedied) [reply libs/langchain-community/src/vectorstores/azure_aisearch.ts:375]: It's needed to implement the `maxMarginalRelevanceSearch()` method.
In the next PR I can however add an option to include it (like some other LC integrations) and make it disabled by default

File: libs/langchain-community/src/vectorstores/azure_aisearch.ts [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+          name: DEFAULT_FIELD_CONTENT,
+          searchable: true,
+          filterable: true,
+          type: "Edm.String",
+        },
+        {
+          name: DEFAULT_FIELD_CONTENT_VECTOR,
+          searchable: true,
+          type: "Collection(Edm.Single)",
+          vectorSearchDimensions: 1536,
```
[the remark this answers is not in the document]
Comment (sinedied) [reply libs/langchain-community/src/vectorstores/azure_aisearch.ts:613]: Good catch. I'll check if we can get it from the LC embedding model directly or add it to the config

File: libs/langchain-community/src/vectorstores/azure_aisearch.ts [hunk 7]
```diff
@@ -0,0 +1,694 @@
...
+          vectorSearchProfileName: "vector-search-profile",
+        },
+        {
+          name: DEFAULT_FIELD_METADATA,
+          type: "Edm.ComplexType",
```
[the remark this answers is not in the document]
Comment (sinedied) [reply libs/langchain-community/src/vectorstores/azure_aisearch.ts:618]: Yes, I'm not quite happy with that one and it comes from the base implementation I started from. 
But I thought as a default as least it makes it easier for the users who do not want to bother with the schema, as LC allows for arbitrary metadata.

That's also why I said I wand to provide more customizations options in another PR to allow for custom index schema for the users who wants better control over the data and filters.

File: yarn.lock [hunk 8]
```diff
@@ -23807,6 +23900,7 @@ __metadata:
     "@aws-sdk/client-sagemaker-runtime": ^3.310.0
     "@aws-sdk/client-sfn": ^3.310.0
     "@aws-sdk/credential-provider-node": ^3.388.0
+    "@aws-sdk/signature-v4": ^3.374.0
```
[the remark this answers is not in the document]
Comment (sinedied) [reply yarn.lock:23903]: Yes, cleaned

File: libs/langchain-community/src/vectorstores/azure_aisearch.ts [hunk 9]
```diff
@@ -0,0 +1,694 @@
...
+            fields: [DEFAULT_FIELD_CONTENT_VECTOR],
+          },
+        ],
+      },
+      filter,
+      top: k,
+      queryType: "semantic",
+      semanticSearchOptions: {
+        configurationName: "semantic-search-config",
+        captions: {
```
[the remark this answers is not in the document]
Comment (sinedied) [reply libs/langchain-community/src/vectorstores/azure_aisearch.ts]: Currently it's not, so I'm removing it. We might always add it later as an option

File: docs/api_refs/typedoc.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (sinedied) [reply docs/api_refs/typedoc.json]: Solved after a rebase from main

File: docs/api_refs/typedoc.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (sinedied) [reply docs/api_refs/typedoc.json]: Solved after a rebase from main

**Comment by sinedied** [comment]
Thanks for the feedbacks! I'm leaving some of the comment open here for reference, as I'd like to tackle these in a separate PR (optional embeddings, improved schema and maybe think of the 100k limit).

It seems there's a still build issue coming from the main branch, but otherwise the PR is ready for the next step.

**Review by pablocastro — APPROVED** [review]
LGTM 

(I think it's fine to defer > 100K docs operations and the metadata structure topic, this is already useful in most scenarios without either of those)

**Review by farzad528 — APPROVED** [review]

**Comment by sinedied** [comment]
Thanks @jacoblee93 for your review and comments, I'll update the PR in the coming days and rebase it from main to fix the conflict

File: libs/langchain-community/src/vectorstores/azure_aisearch.ts [hunk 10]
```diff
@@ -0,0 +1,732 @@
...
+    this.options = config.search;
+    this.embeddings = embeddings;
+  }
+
+  /**
+   * Removes specified documents from the AzureAISearchVectorStore using a filter.
+   * @param filter OData filter to find documents to delete.
+   * @returns A promise that resolves when the documents have been removed.
+   */
+  async deleteMany(filter: string): Promise<IndexingResult[]> {
```
[the remark this answers is not in the document]
Comment (sinedied) [reply libs/langchain-community/src/vectorstores/azure_aisearch.ts]: Thanks, fixed

File: libs/langchain-community/src/vectorstores/azure_aisearch.ts [hunk 11]
```diff
@@ -0,0 +1,732 @@
...
+    documents: Document[],
+    options?: AzureAISearchAddDocumentsOptions
+  ) {
+    const texts = documents.map(({ pageContent }) => pageContent);
+    const results: string[] = [];
+
+    for (let i = 0; i < texts.length; i += this.embeddingBatchSize) {
+      const batch = texts.slice(i, i + this.embeddingBatchSize);
+      const docsBatch = documents.slice(i, i + this.embeddingBatchSize);
+      const batchEmbeddings: number[][] = await this.embeddings.embedDocuments(
```
[the remark this answers is not in the document]
Comment (sinedied) [reply libs/langchain-community/src/vectorstores/azure_aisearch.ts:281]: Fixed

File: libs/langchain-community/src/vectorstores/azure_aisearch.ts [hunk 12]
```diff
@@ -0,0 +1,732 @@
...
+    const docsWithScore: [Document, number][] = [];
+
+    for await (const item of results) {
+      const document = new Document<
+        AzureAISearchDocumentMetadata & { embedding: number[] }
+      >({
+        pageContent: item.document[DEFAULT_FIELD_CONTENT],
+        metadata: {
+          ...item.document[DEFAULT_FIELD_METADATA],
+          embedding: item.document[DEFAULT_FIELD_CONTENT_VECTOR],
```
[the remark this answers is not in the document]
Comment (sinedied) [reply libs/langchain-community/src/vectorstores/azure_aisearch.ts:538]: I made it disabled by default, and added the property in the metadata type to surface it clearly, so the prefix should not be needed now

**Comment by sinedied** [comment]
@jacoblee93 I've process all comments an rebased the branch

**Comment by jacoblee93** [comment]
Looks good - unfortunate about metadata filtering but I think better to get it in at this point.

**Comment by jacoblee93** [comment]
Thank you!

**Pull request merged**